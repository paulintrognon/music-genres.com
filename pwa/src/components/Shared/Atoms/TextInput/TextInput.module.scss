.input {
  width: 100%;
  margin: auto;
  height: 3rem;
  color: var(--color-primary);
  background-color: var(--color-secondary);
  padding: 0 1.5rem;
  box-shadow: 0 0.5rem 0.5rem 0 rgba(0,0,0,.2);
  transition: height .2s, font-size .2s;
  box-sizing: border-box;
  font-family: Oxygen, sans-serif;

  &.dirty {
    height: 4rem;
    font-size: var(--font-size-18);
  }
}

