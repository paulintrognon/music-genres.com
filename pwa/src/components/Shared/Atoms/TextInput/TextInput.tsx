import classnames from 'classnames'
import { useRef, useState } from 'react'
import styles from './TextInput.module.scss'

export interface Props extends React.ComponentProps<'input'> {
  /**
   * Called on focus or if text is entered
   */
  onDirty?: () => void

  /**
   * Called on blur *only if input value is empty*
   */
  onClean?: () => void
}
const TextInput: React.FC<Props> = ({ className, onDirty, onClean, onFocus, onBlur, ...props }) => {
  const ref = useRef<HTMLInputElement>(null)

  const [isFocused, setIsFocused] = useState(false)

  /**
   * On Focus, we update the isFocused state, and we call onDirty
   */
  const handleOnFocus = (event: React.FocusEvent<HTMLInputElement>): void => {
    setIsFocused(true)
    onDirty && onDirty()
    onFocus && onFocus(event)
  }

  /**
   * On Blur, we update the isFocused state, and *if the input is empty*, we call onClean
   */
  const handleOnBlur = (event: React.FocusEvent<HTMLInputElement>): void => {
    setIsFocused(false)
    // onClean is called only when blurring from an empty input
    if (onClean && !ref.current?.value) {
      onClean()
    }
    onBlur && onBlur(event)
  }

  /**
   * The input is dirty if focus, or if its value is non-empty.
   */
  const isDirty = isFocused || ref.current?.value

  return (
    <input
      type="text"
      ref={ref}
      className={classnames(className, styles.input, { [styles.dirty]: isDirty })}
      onFocus={handleOnFocus}
      onBlur={handleOnBlur}
      {...props}
    />
  )
}
export default TextInput
