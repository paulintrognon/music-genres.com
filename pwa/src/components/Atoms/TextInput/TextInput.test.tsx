import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import TextInput from './TextInput'

it('should render the TextInput', () => {
  expect(renderer.create(<TextInput placeholder="placeholder" />).toJSON()).toMatchSnapshot()
})

it('should call onDirty when focus, and onClean when blur', () => {
  const onDirty = jest.fn()
  const onClean = jest.fn()
  const component = mount(<TextInput onDirty={onDirty} onClean={onClean} />)

  component.find('input').simulate('focus')
  expect(onDirty).toHaveBeenCalledTimes(1)
  expect(onClean).toHaveBeenCalledTimes(0)

  component.find('input').simulate('blur')
  expect(onDirty).toHaveBeenCalledTimes(1)
  expect(onClean).toHaveBeenCalledTimes(1)
})

it('should NOT call onClean when input as value on blur', () => {
  const onClean = jest.fn()
  const component = mount(<TextInput onClean={onClean} />)

  component.find('input').simulate('focus')
  component.find('input').getDOMNode<HTMLInputElement>().value = 'new value'
  component.find('input').simulate('blur')
  expect(onClean).toHaveBeenCalledTimes(0)
})

it('should call parent onFocus & onBlur', () => {
  const onFocus = jest.fn()
  const onBlur = jest.fn()
  const component = mount(<TextInput onFocus={onFocus} onBlur={onBlur} />)
  expect(onFocus).toHaveBeenCalledTimes(0)
  expect(onBlur).toHaveBeenCalledTimes(0)

  component.find('input').simulate('focus')
  expect(onFocus).toHaveBeenCalledTimes(1)
  expect(onBlur).toHaveBeenCalledTimes(0)

  component.find('input').simulate('blur')
  expect(onFocus).toHaveBeenCalledTimes(1)
  expect(onBlur).toHaveBeenCalledTimes(1)
})

it('onDirty & onClean should still work if we provide onFocus & onBlur', () => {
  const onFocus = jest.fn()
  const onBlur = jest.fn()
  const onDirty = jest.fn()
  const onClean = jest.fn()
  const component = mount(
    <TextInput onFocus={onFocus} onBlur={onBlur} onDirty={onDirty} onClean={onClean} />
  )
  expect(onFocus).toHaveBeenCalledTimes(0)
  expect(onBlur).toHaveBeenCalledTimes(0)
  expect(onDirty).toHaveBeenCalledTimes(0)
  expect(onClean).toHaveBeenCalledTimes(0)

  component.find('input').simulate('focus')
  expect(onFocus).toHaveBeenCalledTimes(1)
  expect(onBlur).toHaveBeenCalledTimes(0)
  expect(onDirty).toHaveBeenCalledTimes(1)
  expect(onClean).toHaveBeenCalledTimes(0)

  component.find('input').simulate('blur')
  expect(onFocus).toHaveBeenCalledTimes(1)
  expect(onBlur).toHaveBeenCalledTimes(1)
  expect(onDirty).toHaveBeenCalledTimes(1)
  expect(onClean).toHaveBeenCalledTimes(1)
})
