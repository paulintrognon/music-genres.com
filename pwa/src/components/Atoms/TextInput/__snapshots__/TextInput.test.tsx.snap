// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the TextInput 1`] = `
<input
  className="input"
  onBlur={[Function]}
  onFocus={[Function]}
  placeholder="placeholder"
  type="text"
/>
`;
